### PR TITLE
 refactor: revamp session manaagement module unit test cases

### PR DESCRIPTION
**Pull request by SKannaniOS**
## Description
This pull request comprehensively migrates unit test cases from XCTest framework to the modern Swift Testing framework for session management and lifecycle management modules. The migration includes refactoring test structure, enhancing test coverage with parameterized tests, and improving test readability and maintainability.

The changes are initiative to revamp unit test cases for lifecycle management, transitioning to a more modern, expressive, and efficient testing approach.

## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [x] Code refactor/optimization

### Session Management Tests Migration
- **SessionActionTests.swift**: Complete rewrite from XCTest to Swift Testing
  - Migrated from class-based `XCTestCase` to struct-based `@Suite`
  - Replaced XCTAssert statements with modern `#expect` assertions
  - Added comprehensive parameterized tests for all session actions:
    - `UpdateSessionIdAction` with various ID scenarios (0, UInt64.max, standard values)
    - `UpdateIsSessionStartAction` with boolean flag combinations
    - `UpdateSessionTypeAction` with manual/automatic type transitions
    - `UpdateSessionLastActivityAction` with activity time edge cases
    - `EndSessionAction` with comprehensive state reset validation
  - Enhanced immutability testing to ensure Redux-like state management integrity
  - Added descriptive test case structures for better test organization

- **SessionHandlerTests.swift**: Complete modernization and restructuring
  - Migrated from XCTest to Swift Testing framework
  - Removed platform-specific conditional compilation for notification handling
  - Replaced verbose given-when-then structure with concise, descriptive test methods
  - Added comprehensive parameterized tests covering:
    - Session initialization with various configurations
    - Session management operations (start, end, refresh)
    - Automatic session behavior and transitions
    - Session timeout calculations and validation
    - Session state persistence across handler instances
    - Session type transitions (manual ↔ automatic)
  - Fixed edge case handling for session ID 0 (treated as invalid)
  - Enhanced integration testing for complete session lifecycle

- **SessionInfoTests.swift**: Streamlined test structure and coverage
  - Migrated from XCTest class to Swift Testing suite
  - Simplified setup using modern initialization patterns
  - Added parameterized tests for storage operations
  - Enhanced test coverage for invalid data handling
  - Added comprehensive integration tests for session lifecycle
  - Improved test organization with clear section markers

### Key Technical Improvements
- **Parameterized Testing**: Extensive use of arguments in `@Test` annotations for comprehensive scenario coverage
- **Type Safety**: Leveraged Swift's type system for better test data organization
- **Performance**: Eliminated unnecessary setup/teardown overhead through struct-based test organization
- **Readability**: Clear, descriptive test names following Swift Testing conventions
- **Maintainability**: Consolidated test logic and reduced code duplication

## Checklist
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added the necessary documentation (if appropriate).
- [x] I have ensured that my code follows the project's code style.
- [x] I have checked for potential performance impacts and optimized if necessary.
- [x] I have checked the code for security issues.
- [ ] I have updated the changelog (if required).

### Prerequisites
- Xcode 15.0+ with Swift Testing framework support
- Run tests using Swift Testing instead of XCTest

### Test Execution
```bash
# Run all session management tests
swift test --filter SessionManagementTests

# Run specific test suites
swift test --filter SessionActionTests
swift test --filter SessionHandlerTests
swift test --filter SessionInfoTests
```

### Test Coverage Verification
The migrated tests provide comprehensive coverage for:
- Session initialization and configuration handling
- Session state management (start, end, refresh operations)
- Automatic vs manual session transitions
- Session timeout behavior and validation
- Session persistence across handler instances
- Redux-like action dispatching and state immutability
- Storage operations and data consistency
- Lifecycle event handling and observer patterns

### Expected Results
- All tests should pass with the new Swift Testing framework
- Test execution should be faster due to reduced setup overhead
- Test output should be more descriptive and easier to debug
- Parameterized tests should provide better coverage of edge cases

## Breaking Changes
This is not a breaking change for the SDK functionality, but it does introduce changes for developers working on the test suite.

## Maintainers Checklist
- [ ] The code has been reviewed.
- [ ] CI tests have passed.
- [ ] All necessary documentation has been updated.

### Related Work
- This PR is part of the broader initiative to modernize the test infrastructure
- Aligns with Swift Testing best practices and modern iOS development standards
- Improves test maintainability and developer experience

### Benefits of Swift Testing Migration
1. **Better Performance**: Reduced test execution time through optimized framework
2. **Enhanced Readability**: More expressive assertions and cleaner syntax
3. **Improved Debugging**: Better error messages and test failure context
4. **Modern Patterns**: Leverages latest Swift language features
5. **Parameterized Testing**: Built-in support for data-driven tests
6. **Type Safety**: Compile-time validation of test data and expectations

### Files Modified
- `Tests/RudderStackAnalyticsTests/SessionManagementTests/SessionActionTests.swift` - Complete rewrite
- `Tests/RudderStackAnalyticsTests/SessionManagementTests/SessionHandlerTests.swift` - Complete modernization
- `Tests/RudderStackAnalyticsTests/SessionManagementTests/SessionInfoTests.swift` - Framework migration

### Test Statistics
- **Total Test Methods**: 25+ comprehensive test methods across all suites
- **Parameterized Test Cases**: 50+ different parameter combinations
- **Test Coverage**: Enhanced coverage for edge cases and integration scenarios
- **Code Reduction**: ~40% reduction in test boilerplate code

### CODE REVIEW

File: Tests/RudderStackAnalyticsTests/SessionManagementTests/SessionHandlerTests.swift [hunk 1]
```diff
@@ -5,239 +5,287 @@
...
+        let sessionHandler = SessionHandler(analytics: analytics)
+        
+        switch configuration.automaticSessionTracking {
+        case true:
+            #expect(sessionHandler.sessionId != nil)
+            #expect(sessionHandler.isSessionStart != SessionConstants.defaultIsSessionStart)
+        case false:
+            #expect(sessionHandler.sessionId == nil)
+            #expect(sessionHandler.isSessionStart == SessionConstants.defaultIsSessionStart)
+        }
```
Comment (1abhishekpandey) [inline Tests/RudderStackAnalyticsTests/SessionManagementTests/SessionHandlerTests.swift]: Do we really need the `false` case?

Let's try to avoid using `switch` case in tests.

File: Tests/RudderStackAnalyticsTests/SessionManagementTests/SessionHandlerTests.swift [hunk 2]
```diff
@@ -5,239 +5,287 @@
...
+        sessionHandler.refreshSession()
+        
+        switch hasActiveSession {
+        case true:
+            #expect(sessionHandler.sessionId != originalSessionId, "Session ID should be refreshed")
+            #expect(sessionHandler.sessionId != nil, "Session ID should not be nil after refresh")
+            #expect(sessionHandler.isSessionStart, "Session should be marked as started")
+            #expect(sessionHandler.sessionType == .manual, "Session type should remain manual")
+        case false:
+            #expect(sessionHandler.sessionId == nil, "Session ID should remain nil when no active session exists")
```
Comment (1abhishekpandey) [inline Tests/RudderStackAnalyticsTests/SessionManagementTests/SessionHandlerTests.swift]: Same over here.

File: Tests/RudderStackAnalyticsTests/SessionManagementTests/SessionHandlerTests.swift [hunk 3]
```diff
@@ -5,239 +5,287 @@
...
-                then("A new session should be started if timed out") {
-                    XCTAssertTrue(manager.isSessionTimedOut, "Session should be timed out")
-                    XCTAssertNotEqual(manager.sessionId, initialSessionId, "A new session should be created after timeout")
-                }
+        } else {
+            if existingSessionType == .automatic {
+                #expect(sessionHandler.sessionId == nil)
+            } else if existingSessionType == .manual {
+                #expect(sessionHandler.sessionId == originalSessionId)
+                #expect(sessionHandler.sessionType == .manual)
```
Comment (1abhishekpandey) [inline Tests/RudderStackAnalyticsTests/SessionManagementTests/SessionHandlerTests.swift]: Can we simplify this, i.e., remove nested if-else-if checks for the assert phase?

File: Tests/RudderStackAnalyticsTests/SessionManagementTests/SessionInfoTests.swift [hunk 4]
```diff
@@ -6,60 +6,132 @@
...
+        mockStorage.write(value: invalidValue, key: storageKey)
+        sessionInfo = SessionInfo.initializeState(mockStorage)
+        
+        switch fieldName {
+        case "sessionId":
+            #expect(sessionInfo.id == SessionConstants.defaultSessionId)
+        case "lastActivityTime":
+            #expect(sessionInfo.lastActivityTime == SessionConstants.defaultSessionLastActivityTime)
+        default:
+            break
```
Comment (1abhishekpandey) [inline Tests/RudderStackAnalyticsTests/SessionManagementTests/SessionInfoTests.swift]: Remove switch statement.

File: Tests/RudderStackAnalyticsTests/SessionManagementTests/SessionInfoTests.swift [hunk 5]
```diff
@@ -6,60 +6,132 @@
...
+        sessionInfo.storeIsSessionStart(isSessionStart: isSessionStart, storage: mockStorage)
+        
+        let storedValue: Bool? = mockStorage.read(key: Constants.storageKeys.isSessionStart)
+        #expect(storedValue == isSessionStart)
+    }
+
+    @Test("when storing session type, then it is saved correctly", arguments: [
+        (SessionType.manual, true),
+        (SessionType.automatic, false)
+    ])
```
Comment (1abhishekpandey) [inline Tests/RudderStackAnalyticsTests/SessionManagementTests/SessionInfoTests.swift]: Can we write the given phase?

File: Tests/RudderStackAnalyticsTests/SessionManagementTests/SessionInfoTests.swift [hunk 6]
```diff
@@ -6,60 +6,132 @@
...
-                then("SessionInfo initialized with default values") {
-                    XCTAssertTrue(sessionInfo.id == SessionConstants.defaultSessionId)
-                    XCTAssertTrue(sessionInfo.isStart == SessionConstants.defaultIsSessionStart)
-                    XCTAssertTrue(sessionInfo.type == SessionConstants.defaultSessionType)
-                }
-            }
+    @Test("when initialized with invalid data, then default values are used", arguments: [
+        ("sessionId", Constants.storageKeys.sessionId, "invalid_number"),
+        ("lastActivityTime", Constants.storageKeys.lastActivityTime, "invalid_time")
+    ])
```
Comment (1abhishekpandey) [inline Tests/RudderStackAnalyticsTests/SessionManagementTests/SessionInfoTests.swift]: Can we write the given phase?

File: Tests/RudderStackAnalyticsTests/SessionManagementTests/SessionHandlerTests.swift (hunk 1 above)
Comment (1abhishekpandey) [reply Tests/RudderStackAnalyticsTests/SessionManagementTests/SessionHandlerTests.swift]: Let's not use the `if-else` or `switch` case in the test to assert the logic.

Ideally, I believe, tests should have assert logic without any `if-else` or `switch` checks. This makes the test simple to read and maintain.

Let's try to write different test cases for if and else separately, if needed.

Objective:
- We should try to avoid writing `if-else` or `switch` in the assert phase, as it makes the test hard to read and maintain.

Comment (1abhishekpandey) [reply Tests/RudderStackAnalyticsTests/SessionManagementTests/SessionHandlerTests.swift]: I'll wait for you to refactor all the test where you've used if-else checks in the assert phase.

Comment (SKannaniOS) [reply Tests/RudderStackAnalyticsTests/SessionManagementTests/SessionHandlerTests.swift]: The issue is that this test case is parameterized, so using a switch or if-else statement is necessary to handle the different parameter values.

Comment (1abhishekpandey) [reply Tests/RudderStackAnalyticsTests/SessionManagementTests/SessionHandlerTests.swift]: Let's not use parameterised test, if that's the case!

We should use parameterised tests in such a way that no `if-else` need to be used. Use parameterised tests when all cases follow the same flow and only the inputs/outputs change.

Let's refactor the code.

**Review by 1abhishekpandey — APPROVED** [review]
